README.rst
Baka Framework
==============

|docs| |python| |pypi| |license|

`Baka Framework`_ adalah web application framework yang menggunakan core wsgi dari Pyramid.


Penggunaan
==========

Kamu dapat menggunakan baka framework dengan sangat sederhana seperti ``route handler function``, misalnya.

.. code:: python

   from baka import Baka
   from baka.log import log

   app = Baka(__name__)

   # route method
   @app.route('/')
   def index_page(req):
       log.info(req)
       return {'Baka': 'Hello World!'}


   @app.route('/home')
   def home_page(req):
       log.info(req)
       return {'Route': 'home'}


   # root resources routes
   class ResourcesPage(object):
       def __init__(self, request):
           self._name = 'Resource Page'
           log.info(request.params)


   # GET resource method
   @ResourcesPage.GET()
   def resources_page_get(root, request):
       return {
          'hello': 'Get Hello resources from Page root %s ' % page._name
       }


Modular Package / Folder
========================

Dengan penggunakan ``baka.include(callable)``, kamu dapat menggabungkan module terpisah dari beberapa file didalam *package module*.

``contoh file: testbaka/view_user.py``


.. code:: python

   from .app import app

   @app.route('/users')
   def user(req):
       return {'users': 'all data'}

   def includeme(config):
       pass


``file: testbaka/app.py``


.. code:: python

   from baka import Baka
   from baka.log import log

   app = Baka(__name__)
   app.include('testbaka.view_user') # include module dari file view_user.py


   @app.route('/')
   def index_page(req):
       log.info(req)
       return {'Baka': 'Hello World!'}


   @app.route('/home')
   def home_page(req):
       log.info(req)
       return {'Route': 'home'}


App Folder
==========

Untuk Struktur Application Folder ``optional``

.. code:: yaml

   root
      package (AppBaka)
         config ``optional, Baka(__name__, config_schema=True)``
            config.yaml # digunakan for baka default configuration
         __init__.py # the code goes in here
         wsgi.py # for running in wsgi container e.g gunicorn
      run.py # running development server


Default Configuration Baka from ``config.yaml``

.. code-block:: yaml

   package: AppBaka # mandatory for root package
   version: 0.1.0 # optional
   baka:
       debug_all: True # mandatory for debug environment
       meta:
           version: 0.1.0 # mandatory for json response version


WSGI Container Application Server ``wsgi.py``

.. code:: python

   # -*- coding: utf-8 -*-
   """
       WSGI Application Server
       ~~~~~~~~~
       :author: nanang.jobs@gmail.com
       :copyright: (c) 2017 by Nanang Suryadi.
       :license: BSD, see LICENSE for more details.

       wsgi.py
   """
   from . import app

   application = app


Running in Development mode ``run.py``

.. code:: python

   # -*- coding: utf-8 -*-
   """
       :author: nanang.jobs@gmail.com
       :copyright: (c) 2017 by Nanang Suryadi.
       :license: BSD, see LICENSE for more details.

       run.py.py
   """
   from . import app

   app.run(use_reloader=True)


Install
=======

.. code:: python

   pip install baka


Running
=======

Development mode

.. code::

   python run.py


Production mode with Gunicorn

.. code::

   gunicorn -w 1 -b 0.0.0.0:5000 AppBaka.wsgi


Contoh Aplikasi
===============

.. code::

   git clone https://github.com/baka-framework/baka.git

   cd examples

   python3 -m venv env

   source env/bin/active

   pip install baka

   python run.py


Saran dan Kontribusi
====================

    Qoutes from heroes.

    “ Learning without thinking is useless, but thinking without learning is very dangerous! ”

    -― Sukarno, Di Bawah Bendera Revolusi : Jilid 1

    “ Apabila dalam diri seseorang masih ada rasa malu dan takut untuk berbuat suatu kebaikan, maka jaminan bagi orang tersebut adalah tidak akan bertemunya ia dengan kemajuan selangkah pun ”

    -- Sukarno

    “ Kurang cerdas dapat diperbaiki dengan belajar, kurang cakap dapat dihilangkan dengan pengalaman. Namun tidak jujur sulit diperbaiki. ”

    -- Bung Hatta

    “ Keberanian bukan berarti tidak takut, keberanian berarti menaklukan ketakutan. ”

    -- Bung Hatta


.. |license| image:: https://img.shields.io/pypi/l/baka.svg
    :alt: License
    :target: https://pypi.python.org/pypi/baka/0.4.2

.. |pypi| image:: https://img.shields.io/pypi/v/baka.svg
    :alt: Pypi package manager
    :target: https://pypi.python.org/pypi/baka/0.4.2

.. |python| image:: https://img.shields.io/pypi/pyversions/baka.svg
    :alt: Python version
    :target: https://pypi.python.org/pypi/baka/0.4.2

.. |docs| image:: https://readthedocs.org/projects/baka-framework/badge/?version=latest
    :alt: Documentation Status
    :target: https://baka-framework.readthedocs.io/en/latest/?badge=latest

.. |docs_dev| image:: https://readthedocs.org/projects/baka-framework/badge/?version=develop
    :alt: Documentation Status
    :target: https://baka-framework.readthedocs.io/en/latest/?badge=develop


CHANGELOG.rst
Baka Changelog
===============

Di sini Anda bisa melihat daftar lengkap perubahan di antara setiap rilis Baka framework.



Version 0.4
------------

Major release, unreleased codename ``aiko``

- add docs, `Bakadocs`_ untuk memudahkan cara menggunakan.
- Remove ``app.scan()``, tidak perlu menggunakan fitur scan dari venusian.
- Dependency less venusian, ``venusian.attach(wrapped, callback, 'pyramid', depth=depth)`` menghilangkan dependency venusian di feature route.
- add pyramid ``authenticate and authorization`` policy settings.
- add error handler function, ``error_handler(code)`` route, dengan memasukan http code untuk me-redirect ke handler yang sudah di buat.
- add exception handler function, ``exception_handler`` route, default code 509 dengan pesan status berisi error exception `(resiliency)` methodologi.
- add ``trafaret`` validator, default False untuk `config_schema` values.
- add directive ``add_config_validator``, untuk schema trafaret validator.
- add directive ``get_settings_validator``, untuk menvalidasi trafaret validator.
- custom JSON renderer, ``JSONEncoder``, type object dump (ObjectId untuk mongodb id); uuid type; datetime type; EnumInt Type; enum
- add validator config key, ``validator``, default False, untuk menjalankan fungsi validator dari ``get_settings_validator`` pada setiap baka plugin


.. _Bakadocs: http://baka-framework.readthedocs.io/en/latest/


baka/app.py
# -*- coding: utf-8 -*-
"""
    baka.app
    ~~~~~~~~~
    This module implements the central WSGI application object.

    :copyright: (c) 2016 by Nanang Suryadi.
    :license: BSD, see LICENSE for more details.
"""
import logging
import logging.config
import os
import sys

from pyramid import httpexceptions
from pyramid.config import Configurator
from pyramid.exceptions import NotFound
from pyramid.path import DottedNameResolver
from pyramid.security import NO_PERMISSION_REQUIRED
from pyramid.session import SignedCookieSessionFactory
from pyramid.view import AppendSlashNotFoundViewFactory

from baka._compat import text_type
from .config import trafaret_yaml
from .log import log, _logging_format
from .resources import METHODS, ViewDecorator, default_options_view, unsupported_method_view, _BakaExtensions
from .routes import add_simple_route
from .settings import SettingError


class Baka(object):

    def __init__(self, package,
                 session_key='sekret',
                 authn_policy=None, authz_policy=None,
                 **settings):
        """initial config for singleton baka framework

        :param import_name: the name of the application package
        :param session_key: secret key for session
        :param authn_policy: authenticate policy middleware function
        :param authz_policy: authorization policy middleware function
        :param config_schema: boolean value for trafaret validator
        :param settings: *optional dict settings for pyramid configuration
        """
        self.package = package
        session_factory = SignedCookieSessionFactory(session_key)
        settings.update({
            'secret_key': session_key,
            'session_factory': session_factory,
            'authentication_policy': authn_policy,
            'authorization_policy': authz_policy
        })

        self.config = self.configure(settings)
        self.config.registry.package = self.package
        log.info(self.config.registry)
        self.config.begin()
        self.config.add_view(AppendSlashNotFoundViewFactory(), context=NotFound)
        self.config.add_directive('add_ext', self.add_ext_config)

        self.config.registry.trafaret = None
        self.config.include(__name__)
        self.registry = _BakaExtensions
        _logging_format(self.config.get_settings())
        self.config.commit()
        print('🚀 started: Baka Framework')

    def configure(self, settings):
        """ This initial settings of pyramid Configurator
        :param settings: :dict settings of Configurator
        :return: pyramid.config.Configurator
        """
        if settings.get('environ') is None:
            environ = os.environ
        if settings is None:
            settings = {}

        if settings.get('env') is not None:
            for s in settings.get('env'):
                try:
                    result = s(environ)
                except SettingError as e:
                    log.warn(e)

                if result is not None:
                    settings.update(result)

        if 'secret_key' not in settings:
            log.warn('No secret key provided: using transient key. Please '
                     'configure the secret_key setting or the SECRET_KEY '
                     'environment variable!')
            settings['secret_key'] = os.urandom(64)

        # Set up SQLAlchemy debug logging
        if 'debug_query' in settings:
            level = logging.INFO
            if settings['debug_query'] == 'trace':
                level = logging.DEBUG
            logging.getLogger('sqlalchemy.engine').setLevel(level)

        return Configurator(settings=settings)

    def include(self, callable):
        self.config.include(callable)

    @property
    def name(self):
        """The name of the application.  This is usually the import name
        with the difference that it's guessed from the run file if the
        import name is main.  This name is used as a display name when
        Baka needs the name of the application.  It can be set and overridden
        to change the value.
        """

        module = sys.modules[self.package]
        f = getattr(module, '__file__', '')
        if f in ['__init__.py', '__init__$py']:
            # Module is a package
            return module
        # Go up one level to get package
        package_name = module.__name__.rsplit('.', 1)[0]
        return sys.modules[package_name]

    def add_ext_config(self, config, name, directive):
        """ This does the same thing as :meth:`add_ext` but using for pyramid configuration
        :param config: the config from pyramid configuration object
        :param name: the name of directive
        :param directive: the callback function directive
        """
        self.add_ext(name, directive)

    def add_ext(self, name, directive):
        """ Registry of baka directive
        :param name: the name of directive
        :param directive: the callback function directive
        """
        if name in vars(self).keys():
            raise AttributeError(name)

        c = DottedNameResolver().maybe_resolve(directive)
        setattr(self, name, c)

    def resource(self, path, **kwargs):
        def decorator(wrapped):
            route_name = kwargs.pop("route_name", None)
            route_name = route_name or wrapped.__name__
            route_name = kwargs.pop("name", route_name)
            wrapped.route_name = route_name

            for method in METHODS:
                setattr(wrapped, method,
                        type('ViewDecorator%s' % method,
                             (ViewDecorator, object),
                             {'request_method': method,
                              'state': wrapped,
                              'kwargs': kwargs,
                              'config': self.config
                              }))

            self.config.add_route(route_name, path, factory=wrapped)
            self.config.add_view(default_options_view, route_name=route_name,
                                 request_method='OPTIONS', permission=NO_PERMISSION_REQUIRED)
            self.config.add_view(unsupported_method_view, route_name=route_name, renderer='json')

            return wrapped

        return decorator

    def route(self, path, **kwargs):
        """A decorator that is used to register a view function for a
        given URL rule.  This does the same thing as :meth:`add_simple_route`
        but is intended for decorator usage::

            @app.route('/', renderer='todo:templates/index.html')
            def index(req):
                return {'baka', 'Hello World'}

            @app.route('/', renderer='json')
            def index(req):
                return {'baka', 'Hello World'}

            @app.route('/')
            def index(req):
                return {'baka', 'Hello World'}

        ```pyramid.view.view_config```
        A function, class or method :term:`decorator` which allows a
        developer to create view registrations nearer to a :term:`view
        callable` definition than use :term:`imperative
        configuration` to do the same.

        :param rule: the URL rule as string
        :param renderer: (default: json) the format templates for render in response view
        :param kwargs: supports the following keyword arguments:
                        ``context``, ``exception``, ``permission``, ``name``,
                        ``request_type``, ``route_name``, ``request_method``, ``request_param``,
                        ``containment``, ``xhr``, ``accept``, ``header``, ``path_info``,
                        ``custom_predicates``, ``decorator``, ``mapper``, ``http_cache``,
                        ``require_csrf``, ``match_param``, ``check_csrf``, ``physical_path``, and
                        ``view_options``.
        """
        kwargs['path'] = path

        def decorator(wrapped):
            """Attach the decorator with Venusian"""

            log.debug(kwargs.get('request_method', 'GET'))
            log.debug(kwargs.get('route_name', 'route_name'))

            # Default to not appending slash
            if not "append_slash" in kwargs:
                append_slash = False

            add_simple_route(self.config, wrapped, **kwargs)

            return wrapped

        return decorator

    def run(self, host=None, port=None, **options):
        """ application runner server for development stage. not for production.

        :param host: url host application server
        :param port: number of port
        :param options: dict options for werkzeug wsgi server
        """
        self.config.end()
        settings = self.config.get_settings()
        _host = '127.0.0.1'
        _port = 5000

        host = host or _host
        port = int(port or _port)
        options.setdefault('use_reloader', settings.get('debug_all'))
        options.setdefault('use_debugger', settings.get('debug_all'))

        print('🌎  Listening on port {PORT}'.format(PORT=port))
        print('💻 !Important, you are in development mode.')

        from werkzeug.serving import run_simple
        run_simple(host, port, self.config.make_wsgi_app(), **options)

    def wsgi_app(self, env, response):
        """ wsgi application instance
        :param env: environ wsgi
        :param response: response wsgi
        :return: wsgi middleware
        """
        self.config.end()
        app = self.config.make_wsgi_app()
        return app(env, response)

    def __call__(self, env, response):
        """Shortcut for :attr:`wsgi_app`."""
        return self.wsgi_app(env, response)

    def redirect(self, url):
        raise httpexceptions.HTTPFound(location=url)

    def listen(self, event_type=None):
        def decorator(func):
            self.config.add_subscriber(func, event_type)

        return decorator

    def notify(self, event):
        self.config.registry.notify(event)

    def exception_handler(self, **settings):
        def decorator(wrapped):
            settings['renderer'] = settings.get('renderer', 'json')
            settings['permission'] = settings.get('permission,', NO_PERMISSION_REQUIRED)

            target = DottedNameResolver().maybe_resolve(wrapped)

            def err_func(context, request):
                if not isinstance(context, Exception):
                    context = request.exception or context
                request.response.status = '509 {}'.format(text_type(context))
                # request.response.status_code = 500
                return target(context, request)

            self.config.add_view(err_func, context=Exception, **settings)

            return wrapped

        return decorator


    def error_handler(self, code, **settings):
        def decorator(wrapped):
            settings['renderer'] = settings.get('renderer', 'json')
            settings['permission'] = settings.get('permission,', NO_PERMISSION_REQUIRED)
            target = DottedNameResolver().maybe_resolve(wrapped)
            def err_func(context, request):
                if not isinstance(context, Exception):
                    context = request.exception or context
                request.response.status = context.status
                request.response.status_code = code
                return target(context, request)

            exc = type(httpexceptions.exception_response(code))
            if exc is not None:
                view = err_func
                if exc is NotFound:
                    view = AppendSlashNotFoundViewFactory(err_func)
            else:
                exc = Exception

            self.config.add_view(view, context=exc, **settings)

            return wrapped

        return decorator


def includeme(config):
    config.include('.config')
    config.include('.renderers')
    config.include('.routes')
